SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs

```diff
@@ -0,0 +1,172 @@
+                    this.log.Debug(
+                        "Already moved state machine forward, continuing to check for desired property changes",
+                        () => new { this.deviceId });
+                }
+
+            }
+            catch (Exception e)
+            {
+                this.log.Error("UpdateReportedProperties failed",
+                    () => new { this.deviceId, e.Message, Error = e.GetType().FullName});
+            }
+            finally
+            {
+                //TODO: Here we should unpause the timer - this same thing should be done in all state machine methods
+
+            }
+
+        }
+
+
+        public void Setup(string deviceId, DeviceModel deviceModel)
+        {
+
+            if (this.setupDone)
+            {
+                this.log.Error("Setup has already been invoked, are you sharing this instance with multiple devices?",
+                    () => new { this.deviceId });
+                throw new DeviceActorAlreadyInitializedException();
+            }
+
+            this.setupDone = true;
+            this.deviceId = deviceId;
+            this.deviceModel = deviceModel;
+
+        }
+
+        private bool ChangeTwinPropertiesToMatchActorState(Device device, Dictionary<string, object> actorState)
+        {
+            bool differences = false;
+
+            foreach (var item in actorState)
+            {
+                if (device.Twin.ReportedProperties.ContainsKey(item.Key))
+                {
+                    if (device.Twin.ReportedProperties[item.Key].ToString() != actorState[item.Key].ToString())
+                    {
+                        // Update the Hub twin to match the actor state
+                        device.Twin.ReportedProperties[item.Key] = actorState[item.Key].ToString();
+                        differences = true;
+                    }
+                }
+            }
+            return differences;
+        }
+
+        private bool ChangeTwinPropertiesToMatchDesired(Device device, Dictionary<string, object> actorState)
+        {
+            bool differences = false;
+
+            foreach (var item in device.Twin.DesiredProperties)
+            {
+                if (device.Twin.ReportedProperties.ContainsKey(item.Key))
+                {
+                    if (device.Twin.ReportedProperties[item.Key].ToString() != device.Twin.DesiredProperties[item.Key].ToString())
+                    {
+                        // update the hub reported property to match match hub desired property
+                        device.Twin.ReportedProperties[item.Key] = device.Twin.DesiredProperties[item.Key];
+
+                        // update actor state property to match hub desired changes
+                        if (actorState.ContainsKey(item.Key))
+                            actorState[item.Key] = device.Twin.DesiredProperties[item.Key];
+
+                        differences = true;
+                    }
+                }
```

Comment:
also add new keys

Reply:
I thought about this and I don't want to make the change right now.  I'm leery of letting the simulator be affected by people randomly adding desired properties.  I'll revisit if we have a use case.